### ISSUE TITLE
Error when calling DTr

### ISSUE THREAD

**Issue by Hackfl3isch** [gap-packages/qpa#98]
When i try to call DTr i get the following error message:
```
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `NullspaceMat' on 1 arguments at /opt/gap-4.13.1/lib/methsel2.g:250 called from
NullspaceMat( matrix ) at /opt/gap-4.13.1/pkg/qpa/lib/functors.gi:152 called from
TransposeOfModule( M ) at /opt/gap-4.13.1/pkg/qpa/lib/functors.gi:327 called from
DTr( i ) at *stdin*:467 called from
<function "TauModules for a representation of a quiver">( <arguments> )
 called from read-eval loop at *stdin*:787
type 'quit;' to quit to outer loop
```
This happens from time to time for no particular reason. Usually after restarting GAP it does not happen again (even when trying it with the same module over the same PathAlgebra). 
What sometimes seems to be the reason is the following: 
In the first line of my code i have "LoadPackage("QPA");" and then 700 lines of functions are loaded in ( in particular a function that calculates the Auslander Reiten Translate for a list of modules). 
When i Load the package myself and then copy in the code after the package is loaded, it seems to work fine.
Every other function seem to be loaded correct.

**Comment by sunnyquiver** [comment]
I don't think that I can say anything without seeing the code that you are running when these errors occur. Could you please supply the code?

**Comment by xHoukakun** [comment]
[Error.txt](https://github.com/user-attachments/files/16835356/Error.txt)


This is the code that is called before the error appears (I removed everything that is called after the function: TauModules; ) 
Since it's a rather long and quite involved program.

The reason why i'm doing this is because i didn't found a way to reproduce this error in a systematic way. It seems to appear random. I can run my program and get solutions (keep GAP open) run it the next day with the active session that before had no problems and get this error. Which makes it a bit weird. 

In this program i calculate all thin modules for a given path algebra and then want to calculate DTr of all those thin modules.

**Comment by sunnyquiver** [comment]
I have run the code in Error.txt, and I am so far unable to produce any error. When you run it, you also run the code after the function TauModules and then use the functions AllThinModules and TauModules again? If so, what is done after the function TauModules would be relevant. I cannot see any thing wrong in the code, but the CALL_WITH_CATCH function I have no experience with and there is very little documentation (as far as I can see) of this function. So I have no idea what it does. What does that function do?

**Comment by xHoukakun** [comment]
Hey this is what i mean with: weird behaviour sometimes it occurs sometimes it doesn't.

Sometimes it occurs when running the code for the first time.
Sometimes it occurs when code has already been ran. 
It really is like someone rolls a dice and decides when the code is throwing an error and when it runs without any error.

What my code basically does afterwards is: Calculate Minimal right approximations of those TauModules, with respect to the subcategory generated by thin modules. It then also writes these functions into a .txt file. 

I'm afraid i don't really get what you mean by "When you run it, you also run the code after the function TauModules and then use the functions AllThinModules and TauModules again?"
I define every function that i use in my program first and then i run these function and sometimes get an error. 
But i sometimes calculate everything and my program is basically finished and then i call TauModules again (without calling AllThinModules) and get an error. So sometimes one can describe it like this: 
1. AllThinModules 
2.  TauModules (Works)
3. Calculates Minimal approximations.
4. Calculate TauModules again for the same thin Modules get the error. 

I cannot emphasize this enough: This occurs really randomly throughout an "idle" phases can play a role in it, it seems. Sometimes it occurs by first calling the function sometimes it occurs after a bit of run-time. 

About that: CALL_WITH_CATCH function you mentioned, what i've tried there is the following: In other Languages one can call a method with some kind of error handling. 
If you look at my program you will notice that this part is where one defines a thin module. This is done by going through every subset of arrows. For a general algebra not every subset defines a module (the relations could be not fitting). This is what i tried to catch there since if the relations aren't fitting it's irrelevant. 

I would not try to look very deep into that function because i have this error occuring even before i used this function to calculate thin modules.  But i haven't been able to reproduce the error myself so i wasn't sure about posting here.

I know this is very frustrating to not have an example when this error occurs or something. It's simply a weird behaviour.

**Comment by sunnyquiver** [comment]
I am loading the file Error.txt via the command Read("Error.txt"); in a GAP session, and then I have not been able to obtain any error message. After having done this I have tried to run the example after the definitions of all functions again and again, but also here no error message. So I am at a loss what is going on. Only thing I can recommend is to check the local variables when you are in the break loop (i.e. when you have the prompt brk>) and see what values they have.  This sometimes would give an idea about what is going on.

**Comment by xHoukakun** [comment]
So after testing a bit further i got the error message again. I still have the break loop on, i'm not that experienced with GAP and error handling. 
The Quiver i used this time was: 
Q:=Quiver( ["v1","v2","v3","v4","v5","v6","v7","v8","v9"], [["v3","v1","a1"],["v3","v2","a2"],["v4","v3","a3"],["v5","v4","a4"],["v5","v6","a5"],["v7","v6","a6"],["v7","v8","a8"],["v9","v8","a9"]] );
A:=PathAlgebra(GF(3),Q);
thin:=AllThinModules(A);
Then on DTr(thin[10]); 
I get the error message.

What i think the root cause is: In Line 125 of functors.gi the function: MinimalGeneratingSetOfModule(M);  is called.
In the break loop it returns: [];
I opened another session of GAP and called that function and it returned something different. 
The reason for this seems to be: TopOfModuleProjection in the MinimalGeneratingSetOfModule function returns the 0 map even if the TopOfModule is non zero. (Both functions return the 0 map or the 0 module).

I took a look at this function and it seems that IncomingArrowsOfVertex(); does not work right since for this session it returns: 
[ [ a1 ], [ a2 ], [ a3, a1, a2 ], [ a4, a3 ], [ a4, a5 ], [ a5, a6 ], [ a6, a8 ], [ a8, a9 ], [ a9 ] ]. 
You can see that this is incorrect since arrows appear multiple times. 

Does this help in any way? I still have the session open and could check a few things if you want.

**Comment by sunnyquiver** [comment]
I am not able to recreate the error that you are getting. Every time a run your example, there is no error occurring. Which version of QPA are you running? 1.34 or 1.35, or any earlier version? I am at a loss what this can be.

**Comment by xHoukakun** [comment]
I'm running version 1.35. 
I'm also clueless on what exactly is happening. I mean it only occurs sometimes and i'm not able to recreate the error on purpose. It also seems that noone else is getting that error so there is a slight chance it's because of something with my PC. 

Normally one should be able to recreate an error like this. But it seems to be not possible, i can't even reproduce the error myself, even if i do the same commands in GAP/QPA as before. 

Note: I had this error in an earlier version as well. I reinstalled GAP already so this seemed not to be the root cause. 

I'm also at loss what it can be, is there maybe documentation on how this function: IncomingArrowsOfVertex(); works? Because from the code i can't "see" how it returns the incomming arrows.

**Comment by sunnyquiver** [comment]
Here is what I get when I enter your quiver and find the incoming arrows for each vertex:

```
gap> Q := Quiver( ["v1","v2","v3","v4","v5","v6","v7","v8","v9"], [["v3","v1","a1"], ["v3","v2","a2"], ["v4","v3","a3"],["v5","v4","a4"], ["v5","v6","a5"], ["v7","v6","a6"], ["v7","v8","a8"], ["v9","v8","a9"]] );
<quiver with 9 vertices and 8 arrows>
gap> List( vertices, IncomingArrowsOfVertex );
[ [ a1 ], [ a2 ], [ a3 ], [ a4 ], [  ], [ a5, a6 ], [  ], [ a8, a9 ], [  ] ]

```
This makes me even more confused. The value of the function IncomingArrowsOfVertex is set when the quiver is constructed. You can see it in the file quiver.gi.

**Comment by xHoukakun** [comment]
Yes i don't understand how the code works but i think it's not relevant.

But here is the reason why you maybe not able to replicate this error:
I defined a quiver Q:="Q1";
Here the error occurred.
Then i overwrote Q with another Quiver. Q:="Q2";
Here everything was correct.
Then i redefined Q :="Q1";
Now everything is correct. 
I think it's not relevant what Q1 and Q2 are exactly they are just quivers. I'm really clueless why this happens, my intuition would suggest that it has something to do how GAP allocates memory? This is just a blind guess. This maybe the reason on why it's difficult to replicate the error as well.

**Comment by xHoukakun** [comment]
Hello after some time this Error popped up on another machine and it got even weirder:  i got a List of quivers (All of Type E_8) and when i ran over that list i calculated DTr (no error appeared) 
I stopped the calculations with strg+c 
Ran over that list to see if really every of those quivers is of type E_8.
then started the first calculations again (Error). Note: the quivers were not changed during the process only the command IsDynkinQuiver were called. 

[Test.txt](https://github.com/user-attachments/files/17773682/Test.txt)

the txt-file is what i have done: In the last few days whenever i checked something with the command: IsDynkinQuiver this error occured. 
Best Regards,

**Comment by fingolfin** [comment]
@Hackfl3isch which version of GAP are you using? On which OS are you? Based on the file I assume Windows?

I am doing this to read your file:
```
gap-4.14.0 ~/Downloads/Error.txt
```
I.e. I start a clean GAP 4.14.0 with your input file and I get no error.

But also if I use `Read` to read your file there is no error.

**Comment by Hackfl3isch** [comment]
@fingolfin  i'm using 4.13.1, but i updated to 4.14.0 and the error still exists when i copied the file into GAP-Terminal. 
I ran the command Read("ErrorT.txt"); with this file:

[ErrorT.txt](https://github.com/user-attachments/files/19629419/ErrorT.txt)

I'm using windows as you assumed, but i think the error also occured on a Linux OS (i'm sure that it happened on my Rasperry Pi as well).

Have you tried reading the Test.txt file i provided https://github.com/gap-packages/qpa/issues/98#issuecomment-2478444909 
As said in that comment, the error occured consistently when running the command IsDynkinQuiver();  My guess is that this function somehow messes up the arrows in a weird way. 

Thank you in advance!

**Comment by sunnyquiver** [comment]
Now I also got the following error message:

```
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `NullspaceMat' on 1 arguments at /Users/oyvinso/gap-4.14.0/lib/methsel2.g:250 called from
NullspaceMat( matrix 
 ) at /Users/oyvinso/gap-4.14.0/pkg/qpa/lib/functors.gi:152 called from
TransposeOfModule( M 
 ) at /Users/oyvinso/gap-4.14.0/pkg/qpa/lib/functors.gi:327 called from
DTr( i ) at ErrorT.txt:181 called from
TauModules( thin ) at ErrorT.txt:254 called from
<function "unknown">( <arguments> )
 called from read-eval loop at ErrorT.txt:254
type 'quit;' to quit to outer loop

```
Investigating this case, I think that I have identified one bug. This is included in the commit 3a456bd. Using this, the commands in ErrorT.txt were executed without errors.

**Comment by Hackfl3isch** [comment]
After doing some tests with your solution, i think there is still another bug hidden somewhere: 

[ErrorArrows.txt](https://github.com/user-attachments/files/19663272/ErrorArrows.txt)

You can see that i overwrote the function TopOfModuleProjection. However the arrows still change after calling IsDynkinQuiver, which may cause other errors at some point in the code. 
For completness i wanted to mention it.

**Comment by sunnyquiver** [comment]
Thank you for the suggestion of an alternative algorithm. But I think that I have indentified the problem with the existing algorithm: `IncomingArrowsOfVertex` is a mutable attribute and this is why its value is changing. However, you avoid this problem by using `ShallowCopy( IncomingArrowsOfVertex( something ) )` instead of just `IncomingArrowsOfVertex( something )`. I have updated the repository to include this update. Clone the repository to obtain the changes.

**Comment by Hackfl3isch** [comment]
Yes i see now, i guess this Issue can be closed now? 
small note:  in Line 265 i think it's suppose to mean: auxillary function :) 
Thank you for fixing this bug
